### PR TITLE
Rename functional functions; Refs #773

### PR DESCRIPTION
**Pull request by mxcl**
Please review carefully since it would suck for our users to have made a mistake *again*.

### CODE REVIEW

File: Sources/Thenable.swift [hunk 1]
```diff
@@ -182,27 +193,29 @@ public extension Thenable where T: Sequence {
         }
     }
 
-    func thenMap<U: Thenable>(on: DispatchQueue? = conf.Q.map, _ transform: @escaping(T.Iterator.Element) throws -> U) -> Promise<[U.T]> {
+    func map<U: Thenable>(on: DispatchQueue? = conf.Q.map, _ transform: @escaping(T.Iterator.Element) throws -> U) -> Promise<[U.T]> {
```
[the remark this answers is not in the document]
Comment (mxcl) [reply Sources/Thenable.swift]: Or maybe the other.

**Comment by stephencelis** [comment]
Theses all look good!

**Comment by mxcl** [comment]
But I'm pretty sure the one I commented on is wrong, is it not?

**Comment by stephencelis** [comment]
Oh dang, lemme give this one last look. My brain skipped right past `Iterator`. That should be `mapValues`.

File: Sources/Thenable.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephencelis) [reply Sources/Thenable.swift]: If I understand this correctly, this is taking a promise that contains a sequence and return a promise that contains a list of promises. I don't know enough about the internals of this lib, but is this any different than `mapValues` above other than the fact that it constrains further on `Thenable`s?

[the remark this answers is not in the document]
Comment (stephencelis) [reply Sources/Thenable.swift]: Oh, I see. It's processing the promises, like a method version of `Promise.all`

[the remark this answers is not in the document]
Comment (mxcl) [reply Sources/Thenable.swift]: Let me create a summary of each function for you. Will add to main discussion.

**Comment by stephencelis** [comment]
Alright, so both that `map` you called out and the `flatMap` below it are definitely doing something more than `map` and `flatMap`. Seems to me:

1. `map`: take a promise that contains a sequence, a transform from each element to another promise, and return a promise that contains an array of the result of all of those promises.
2. `flatMap`: take a promise that contains a sequence, a transform from each element to another promise containing a sequence, and return a promise that contains an array of the result of all of those sequences, flattened.

The first, to me, sounds like a `mapValuesThen`, and the second sounds like a `flatMapValuesThen`. Kinda wordy, but descriptive of the action.

**Comment by mxcl** [comment]
* `Promise<T> => ((T) -> U?) => Promise<U>` (currently flatMap, proposed: compactMap)
* `Promise<[T]> => ((T) -> U) => Promise<[U]>` (currently map, proposed: mapValues)
* `Promise<[T]> => ((T) => Promise<U>)` (currently flatMap, proposed: flatMapValues)
* `Promise<[T]> => ((T) => U?)` (currently compactMap, proposed: compactMapValues)
* `Promise<[T]> => ((T) -> Promise<U>) => Promise<[U]>` (currently thenMap, proposed: map, probably should be: flatMap)
* `Promise<[T]> => ((T) -> Promise<[U]>) => Promise<[U]>` (currently thenFlatMap, proposed: flatMap, probably should be: DUNNO)

**Comment by mxcl** [comment]
K I pushed these changes.

File: Sources/Deprecations.swift [hunk 2]
```diff
@@ -41,3 +41,46 @@ public extension Promise {
...
+    @available(*, deprecated: 6.1, message: "See: `filterValues`")
+    func filter(on: DispatchQueue? = conf.Q.map, test: @escaping (T.Iterator.Element) -> Bool) -> Promise<[T.Iterator.Element]> {
+        return filterValues(on: on, test)
+    }
+}
+
+public extension Thenable where T: Collection {
+    @available(*, deprecated: 6.1, message: "See: `firstValue`")
+    var first: Promise<T.Iterator.Element> {
+        fatalError()
```
Comment (mklbtz) [inline Sources/Deprecations.swift]: Should this be `return firstValue`? You have `last` implemented as `return lastValue`.

Comment (mxcl) [reply Sources/Deprecations.swift]: Good spot

**Comment by nathanhosselton** [comment]
I dislike how wordy so many of the names are becoming, but I do think the names are clear, which is the most important thing. In a perfect world I think we'd have promise-themed names (in the vein of `then`, `firstly`, etc) unique from functional names. But that's a lot of names to come up with, and the functional names are accurate to the operations.

My one tiny gripe is with `compactMap`. It doesn't feel like a `compactMap` since it's not operating on a sequence, and since it continues or fails the promise chain based on whether or not the unwrapped value exists. `compactMap` on sequences doesn't fail. So even though it breaks away from the naming convention we have going here, I do think e.g. `unwrap` is more accurate. This is a minor complaint though.

I'm good with these names.

**Comment by stephencelis** [comment]
> My one tiny gripe is with `compactMap`. It doesn't feel like a `compactMap` since it's not operating on a sequence, and since it continues or fails the promise chain based on whether or not the unwrapped value exists. `compactMap` on sequences doesn't fail.

I think this is a matter of considering what `compactMap` on sequence is doing. It processes _every_ element in a sequence and discards those that are optional. If you ran `compactMap` on a sequence and always return `nil`, you'll get an empty array back, which has zero data you can process: it's failed, in a way. The optional type itself is often used to denote failure. This is a nice thing about looking at things in the abstract. The function signatures are the same so there is some weirdly shared behavior.

Also let's look at `flatMap` on sequences vs. `flatMap` on optionals (which, like promises, either have values or don't): `flatMap` on sequences can change the size of the sequence and feels different than `flatMap` on optionals, which can merely fail over to being `nil`, which in turn feels a _little_ different than `flatMap` on promises, which allows non-failing promises to take their data and make _another_ promise (failing or not). These all feel slightly different, all share the exact same signature, and all allow for a very basic operation: sequencing computations together!

**Comment by mxcl** [comment]
I rebased the branch, hence CI is going again, but it is the same diff as the previous pass. Here is the CI for master:

https://travis-ci.org/mxcl/PromiseKit/builds/343420367

**Pull request merged**